From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sat, 26 Feb 2022 16:43:33 -0800
Subject: [PATCH] Paper PR - Add FillBottleEvents for player and dispenser

https://github.com/PaperMC/Paper/pull/7513

Adds 2 events, PlayerFillBottleEvent and BlockFillBottleEvent. These
events do not have a common superclass due to needing to extend
PlayerEvent and BlockEvent respectively.

TODO: Add a common interface between Block and AreaEffectCloud,
something like BottleSource to add a "source" field to each event to get
the source of the bottle fill, water block, cauldron, beehive, or
AreaEffectCloud.

Co-authored-by: Dmitry Sidorov <jonmagon@gmail.com>

diff --git a/net/minecraft/core/cauldron/CauldronInteractions.java b/net/minecraft/core/cauldron/CauldronInteractions.java
index 04b47a788586d21a37f73ba2dcf9036ea174872b..011c30bb74709ca0dae4f37f2b6bb8d6b626babe 100644
--- a/net/minecraft/core/cauldron/CauldronInteractions.java
+++ b/net/minecraft/core/cauldron/CauldronInteractions.java
@@ -56,7 +56,13 @@ public class CauldronInteractions {
                     }
                     // CraftBukkit end
                     Item usedItem = itemInHand.getItem();
-                    player.setItemInHand(hand, ItemUtils.createFilledResult(itemInHand, player, new ItemStack(Items.GLASS_BOTTLE)));
+                    // Kiterino start - Paper PR - Add FillBottleEvents for player and dispenser
+                    final io.papermc.paper.event.player.PlayerFillBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerFillBottleEvent(player, hand, itemInHand, ItemUtils.createFilledResult(itemInHand, player, new ItemStack(Items.GLASS_BOTTLE)));
+                    if (event.isCancelled()) {
+                        // doesn't seem to require updating the client
+                        return InteractionResult.PASS; // PASS to not increment statistics
+                    }
+                    // Kiterino end - Paper PR - Add FillBottleEvents for player and dispenser
                     player.awardStat(Stats.USE_CAULDRON);
                     player.awardStat(Stats.ITEM_USED.get(usedItem));
                     // level.setBlockAndUpdate(pos, Blocks.WATER_CAULDRON.defaultBlockState()); // CraftBukkit
diff --git a/net/minecraft/core/dispenser/DispenseItemBehavior.java b/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 56b9f9016a652a704d785de683738efcf02a44f9..acc5a81c0193cac07959210f232a1564875c0ed5 100644
--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -523,13 +523,25 @@ public interface DispenseItemBehavior {
                     // Paper end - Call BlockDispenseEvent
                     if (state.is(BlockTags.BEEHIVES, s -> s.hasProperty(BeehiveBlock.HONEY_LEVEL) && s.getBlock() instanceof BeehiveBlock)
                         && state.getValue(BeehiveBlock.HONEY_LEVEL) >= 5) {
+                        // Kiterino start - Paper PR - Add FillBottleEvents for player and dispenser
+                        final io.papermc.paper.event.block.BlockFillBottleEvent bottleEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFillBottleEvent(level, source.pos(), dispensed, new ItemStack(Items.HONEY_BOTTLE));
+                        if (bottleEvent.isCancelled()) {
+                            return dispensed;
+                        }
+                        // Kiterino end - Paper PR - Add FillBottleEvents for player and dispenser
                         ((BeehiveBlock)state.getBlock())
                             .releaseBeesAndResetHoneyLevel(level, state, target, null, BeehiveBlockEntity.BeeReleaseStatus.BEE_RELEASED);
                         this.setSuccess(true);
-                        return this.takeLiquid(source, dispensed, new ItemStack(Items.HONEY_BOTTLE));
+                        return this.takeLiquid(source, dispensed, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(bottleEvent.getResultItem())); // Kiterino - Paper PR - Add FillBottleEvents for player and dispenser
                     } else if (level.getFluidState(target).is(FluidTags.WATER)) {
+                        // Kiterino start - Paper PR - Add FillBottleEvents for player and dispenser
+                        final io.papermc.paper.event.block.BlockFillBottleEvent bottleEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFillBottleEvent(level, source.pos(), dispensed, PotionContents.createItemStack(Items.POTION, Potions.WATER));
+                        if (bottleEvent.isCancelled()) {
+                            return dispensed;
+                        }
+                        // Kiterino end - Paper PR - Add FillBottleEvents for player and dispenser
                         this.setSuccess(true);
-                        return this.takeLiquid(source, dispensed, PotionContents.createItemStack(Items.POTION, Potions.WATER));
+                        return this.takeLiquid(source, dispensed, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(bottleEvent.getResultItem())); // Kiterino - Paper PR - Add FillBottleEvents for player and dispenser
                     } else {
                         return super.execute(source, dispensed);
                     }
diff --git a/net/minecraft/world/item/BottleItem.java b/net/minecraft/world/item/BottleItem.java
index e45fb987b709cbd798b7bcd179a559d8ab70e6ff..a1edc3f09e9112ebbd521e9b41d971ca38edf7d4 100644
--- a/net/minecraft/world/item/BottleItem.java
+++ b/net/minecraft/world/item/BottleItem.java
@@ -33,6 +33,17 @@ public class BottleItem extends Item {
         );
         ItemStack itemStack = player.getItemInHand(hand);
         if (!clouds.isEmpty()) {
+            // Kiterino start - Paper PR - Add FillBottleEvents for player and dispenser
+            final io.papermc.paper.event.player.PlayerFillBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerFillBottleEvent(player, hand, itemStack, new ItemStack(Items.DRAGON_BREATH));
+            if (event.isCancelled()) {
+                player.containerMenu.sendAllDataToRemote(); // client expects the itemstack to be used
+                return InteractionResult.PASS;
+            }
+            final ItemStack resultItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getResultItem());
+            if (resultItem.is(itemStack.getItem())) {
+                player.containerMenu.sendAllDataToRemote(); // client expects the itemstack to be used
+            }
+            // Kiterino end - Paper PR - Add FillBottleEvents for player and dispenser
             AreaEffectCloud cloud = clouds.get(0);
             cloud.setRadius(cloud.getRadius() - 0.5F);
             level.playSound(null, player.getX(), player.getY(), player.getZ(), SoundEvents.BOTTLE_FILL_DRAGONBREATH, SoundSource.NEUTRAL, 1.0F, 1.0F);
@@ -41,7 +52,7 @@ public class BottleItem extends Item {
                 CriteriaTriggers.PLAYER_INTERACTED_WITH_ENTITY.trigger(serverPlayer, itemStack, cloud);
             }
 
-            return InteractionResult.SUCCESS.heldItemTransformedTo(this.turnBottleIntoItem(itemStack, player, new ItemStack(Items.DRAGON_BREATH)));
+            return InteractionResult.SUCCESS.heldItemTransformedTo(this.turnBottleIntoItem(itemStack, player, resultItem)); // Kiterino - Paper PR - Add FillBottleEvents for player and dispenser
         } else {
             BlockHitResult hitResult = getPlayerPOVHitResult(level, player, ClipContext.Fluid.SOURCE_ONLY);
             if (hitResult.getType() == HitResult.Type.MISS) {
@@ -55,10 +66,21 @@ public class BottleItem extends Item {
                 }
 
                 if (level.getFluidState(pos).is(FluidTags.WATER)) {
+                        // Kiterino start - Paper PR - Add FillBottleEvents for player and dispenser
+                        final io.papermc.paper.event.player.PlayerFillBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerFillBottleEvent(player, hand, itemStack, PotionContents.createItemStack(Items.POTION, Potions.WATER));
+                        if (event.isCancelled()) {
+                            player.containerMenu.sendAllDataToRemote(); // client expects the itemstack to be used
+                            return InteractionResult.PASS;
+                        }
+                        final ItemStack resultItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getResultItem());
+                        if (resultItem.is(itemStack.getItem())) {
+                            player.containerMenu.sendAllDataToRemote(); // client expects the itemstack to be used
+                        }
+                        // Kiterino end - Paper PR - Add FillBottleEvents for player and dispenser
                     level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.BOTTLE_FILL, SoundSource.NEUTRAL, 1.0F, 1.0F);
                     level.gameEvent(player, GameEvent.FLUID_PICKUP, pos);
                     return InteractionResult.SUCCESS
-                        .heldItemTransformedTo(this.turnBottleIntoItem(itemStack, player, PotionContents.createItemStack(Items.POTION, Potions.WATER)));
+                            .heldItemTransformedTo(this.turnBottleIntoItem(itemStack, player, resultItem)); // Kiterino - Paper PR - Add FillBottleEvents for player and dispenser
                 }
             }
 
diff --git a/net/minecraft/world/level/block/BeehiveBlock.java b/net/minecraft/world/level/block/BeehiveBlock.java
index 0d3e97012b0d590dc2b93e4886a7009f6f78e0d7..4cc8721b2c81a14286c427bbb33760ff7460a3c5 100644
--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -180,12 +180,25 @@ public class BeehiveBlock extends BaseEntityBlock {
                 hiveEmptied = true;
                 level.gameEvent(player, GameEvent.SHEAR, pos);
             } else if (itemStack.is(Items.GLASS_BOTTLE)) {
+                // Kiterino start - Paper PR - Add FillBottleEvents for player and dispenser
+                final io.papermc.paper.event.player.PlayerFillBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerFillBottleEvent(player, hand, itemStack, new ItemStack(Items.HONEY_BOTTLE));
+                if (event.isCancelled()) {
+                    player.containerMenu.sendAllDataToRemote(); // client expects the itemstack to be used
+                    return InteractionResult.PASS;
+                }
+                final ItemStack resultItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getResultItem());
+                if (resultItem.is(itemStack.getItem())) {
+                    player.containerMenu.sendAllDataToRemote(); // client expects the itemstack to be used
+                }
+                // Kiterino end - Paper PR - Add FillBottleEvents for player and dispenser
                 itemStack.shrink(1);
                 level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.BOTTLE_FILL, SoundSource.BLOCKS, 1.0F, 1.0F);
                 if (itemStack.isEmpty()) {
-                    player.setItemInHand(hand, new ItemStack(Items.HONEY_BOTTLE));
-                } else if (!player.getInventory().add(new ItemStack(Items.HONEY_BOTTLE))) {
-                    player.drop(new ItemStack(Items.HONEY_BOTTLE), false);
+                    // Kiterino start - Paper PR - Add FillBottleEvents for player and dispenser
+                    player.setItemInHand(hand, resultItem);
+                } else if (!player.getInventory().add(resultItem)) {
+                    player.drop(resultItem, false);
+                    // Kiterino end - Paper PR - Add FillBottleEvents for player and dispenser
                 }
 
                 hiveEmptied = true;
